File: TwoPointer/3.longest-substring-without-repeating-characters.go
package leetcode

// @leet start
func valid(freq [128]int) bool {
	for _, v := range freq {
		if v == 2 {
			return false
		}
	}
	return true
}

func lengthOfLongestSubstring(s string) int {
	var (
		l, ret int
		freq   [128]int
	)
	for r, v := range s {
		freq[v]++
		for !valid(freq) {
			freq[s[l]]--
			l++
		}
		ret = max(ret, r-l+1)
	}
	return ret
}

// @leet end

File: TwoPointer/3.longest-substring-without-repeating-characters.py
# @leet start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        maxL = 0
        freq = [0] * 128

        def valid():
            for i in freq:
                if i > 1:
                    return False
            return True

        for r, c in enumerate(s):
            freq[ord(c)] += 1
            while not valid():
                freq[ord(s[l])] -= 1
                l += 1
            maxL = max(maxL, r - l + 1)

        return maxL


# @leet end

